File: sdk/typescript/src/proofs/hashshare.ts
/**
 * HashShare proof (dev tools) — local u-draw & threshold checks.
 *
 * This module helps build and verify HashShare-style proofs client-side for
 * testing and developer tooling. It does NOT perform consensus acceptance; it
 * only reproduces the uniform draw u from (headerHash, nonce[, mixSeed]) and
 * computes S = -ln(u). You can compare S against a micro-nats threshold Θ_µ.
 *
 * Hash domains (by convention here):
 *   - draw domain:   "hashshare.u.v1"
 *   - nullifier dom: "hashshare.nullifier.v1"
 */

import { sha3_256, keccak256 } from '../utils/hash'
import { bytesToHex, hexToBytes, isHex } from '../utils/bytes'
import { encodeCanonicalCBOR, decodeCBOR } from '../utils/cbor'

// ──────────────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────────────

export interface HashShareBody {
  headerHash: string   // 0x… (32 bytes)
  nonce: string        // 0x… (arbitrary length; commonly 8 or 32 bytes)
  mixSeed?: string     // 0x… (optional extra bind)
}

export interface HashShareProof extends HashShareBody {
  /** Deterministic u in (0,1], derived from headerHash/nonce/mixSeed */
  u: number
  /** Score S(u) = -ln(u) in natural units (nats) */
  S: number
  /** Deterministic nullifier for reuse-prevention (dev parity with node) */
  nullifier: string    // 0x…
}

export interface VerifyOptions {
  /** Difficulty threshold in micro-nats; if set, we check S >= Θ (converted to nats). */
  thetaMicro?: number
  /** Choose hash for the draw (defaults to sha3_256). */
  hashFn?: 'sha3_256' | 'keccak256'
}

export interface VerifyResult {
  ok: boolean
  /** Recomputed u (ignores proof.u if present) */
  u: number
  /** Recomputed S in nats */
  S: number
  /** d_ratio = S / Θ (if Θ provided), else null */
  ratio: number | null
  /** True if threshold provided and S >= Θ; else null if no Θ provided */
  meets: boolean | null
  /** Recomputed nullifier */
  nullifier: string
  errors: string[]
}

// Optional envelope helpers (useful when mirroring on-chain formats)
export interface ProofEnvelope {
  typeId: 'HashShare' | number
  body: HashShareBody
  nullifier: string
}

// ──────────────────────────────────────────────────────────────────────────────
// Public API
// ──────────────────────────────────────────────────────────────────────────────

/** Build a local HashShare proof object for tooling/tests. */
export function buildHashShare(body: HashShareBody, opts: VerifyOptions = {}): HashShareProof {
  const headerHash = normalizeHex(body.headerHash)
  const nonce = normalizeHex(body.nonce)
  const mixSeed = body.mixSeed ? normalizeHex(body.mixSeed) : undefined

  const u = drawU(headerHash, nonce, mixSeed, opts.hashFn)
  const S = score(u)
  const nullifier = computeNullifier({ headerHash, nonce, mixSeed })
  return { headerHash, nonce, mixSeed, u, S, nullifier }
}

/** Verify a HashShare proof against optional threshold Θ_µ. */
export function verifyHashShare(bodyOrProof: HashShareBody | HashShareProof, opts: VerifyOptions = {}): VerifyResult {
  const headerHash = normalizeHex(bodyOrProof.headerHash)
  const nonce = normalizeHex(bodyOrProof.nonce)
  const mixSeed = bodyOrProof.mixSeed ? normalizeHex(bodyOrProof.mixSeed) : undefined

  const u = drawU(headerHash, nonce, mixSeed, opts.hashFn)
  const S = score(u)
  const nullifier = computeNullifier({ headerHash, nonce, mixSeed })

  const errors: string[] = []
  // If caller passed a HashShareProof with u/S fields, ensure they match recomputed
  if ('u' in bodyOrProof && typeof bodyOrProof.u === 'number') {
    if (!approxEqual(bodyOrProof.u, u)) errors.push('u mismatch vs recomputed')
  }
  if ('S' in bodyOrProof && typeof bodyOrProof.S === 'number') {
    if (!approxEqual(bodyOrProof.S, S)) errors.push('S mismatch vs recomputed')
  }
  if ('nullifier' in bodyOrProof && typeof (bodyOrProof as any).nullifier === 'string') {
    const given = normalizeHex((bodyOrProof as any).nullifier)
    if (given !== nullifier) errors.push('nullifier mismatch vs recomputed')
  }

  let ratio: number | null = null
  let meets: boolean | null = null
  if (opts.thetaMicro != null) {
    const thetaNats = opts.thetaMicro / 1_000_000
    ratio = thetaNats > 0 ? S / thetaNats : Infinity
    meets = S >= thetaNats
  }

  const ok = errors.length === 0 && (meets !== false)
  return { ok, u, S, ratio, meets, nullifier, errors }
}

/** Convert to a minimal CBOR envelope (typeId + body + nullifier). */
export function toEnvelopeCBOR(p: HashShareProof | HashShareBody): Uint8Array {
  const env: ProofEnvelope = {
    typeId: 'HashShare',
    body: {
      headerHash: normalizeHex(p.headerHash),
      nonce: normalizeHex(p.nonce),
      ...(p.mixSeed ? { mixSeed: normalizeHex(p.mixSeed) } : {})
    },
    nullifier: computeNullifier(p)
  }
  return encodeCanonicalCBOR(env)
}

/** Decode a CBOR envelope previously created by `toEnvelopeCBOR`. */
export function fromEnvelopeCBOR(data: Uint8Array): ProofEnvelope {
  const env = decodeCBOR(data) as ProofEnvelope
  if (!env || typeof env !== 'object') throw new Error('invalid envelope')
  if (!('typeId' in env) || !('body' in env) || !('nullifier' in env)) throw new Error('missing envelope fields')
  env.nullifier = normalizeHex(env.nullifier)
  env.body = {
    headerHash: normalizeHex(env.body.headerHash),
    nonce: normalizeHex(env.body.nonce),
    ...(env.body.mixSeed ? { mixSeed: normalizeHex(env.body.mixSeed) } : {})
  }
  return env
}

// ──────────────────────────────────────────────────────────────────────────────
// Core math & hashing
// ──────────────────────────────────────────────────────────────────────────────

/**
 * Deterministic draw u ∈ (0, 1] from (headerHash, nonce[, mixSeed]).
 * digest = H( "hashshare.u.v1" | headerHash | nonce | mixSeed? )
 * u = (int(digest) + 1) / 2^256   (avoid 0; yields (0,1])
 */
export function drawU(
  headerHash: string,
  nonce: string,
  mixSeed?: string,
  hashFn: VerifyOptions['hashFn'] = 'sha3_256'
): number {
  const dom = text('hashshare.u.v1')
  const hh = hexToBytes(normalizeHex(headerHash))
  const nn = hexToBytes(normalizeHex(nonce))
  const ms = mixSeed ? hexToBytes(normalizeHex(mixSeed)) : new Uint8Array(0)

  const data = concat(dom, hh, nn, ms)
  const digest = hashFn === 'keccak256' ? keccak256(data) : sha3_256(data)

  // Convert 32-byte digest to BigInt (big-endian)
  let acc = 0n
  for (const b of digest) acc = (acc << 8n) | BigInt(b)
  const two256 = 1n << 256n
  const num = acc + 1n
  const u = Number(num) / Number(two256)  // safe enough for uniformity use; we don't need exact 256-bit ratio
  // Clamp to (0,1] against float rounding
  return u > 1 ? 1 : (u <= 0 ? Number.MIN_VALUE : u)
}

/** Score function S(u) = -ln(u) (natural log, in nats). */
export function score(u: number): number {
  if (!(u > 0 && u <= 1)) throw new Error('u must be in (0,1]')
  return -Math.log(u)
}

/**
 * Deterministic nullifier = H("hashshare.nullifier.v1" | headerHash | nonce | mixSeed?)
 * Useful to prevent duplicate submissions of the same HashShare for a header.
 */
export function computeNullifier(p: HashShareBody): string {
  const dom = text('hashshare.nullifier.v1')
  const hh = hexToBytes(normalizeHex(p.headerHash))
  const nn = hexToBytes(normalizeHex(p.nonce))
  const ms = p.mixSeed ? hexToBytes(normalizeHex(p.mixSeed)) : new Uint8Array(0)
  const dig = sha3_256(concat(dom, hh, nn, ms))
  return '0x' + bytesToHex(dig)
}

// ──────────────────────────────────────────────────────────────────────────────
// Utilities
// ──────────────────────────────────────────────────────────────────────────────

function normalizeHex(h: string): string {
  if (typeof h !== 'string') throw new Error('expected hex string')
  const s = h.startsWith('0x') || h.startsWith('0X') ? h : ('0x' + h)
  if (!isHex(s)) throw new Error(`invalid hex: ${h}`)
  return s.toLowerCase()
}

function text(s: string): Uint8Array {
  return new TextEncoder().encode(s)
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const n = parts.reduce((a, p) => a + p.length, 0)
  const out = new Uint8Array(n)
  let off = 0
  for (const p of parts) { out.set(p, off); off += p.length }
  return out
}

function approxEqual(a: number, b: number, eps = 1e-12): boolean {
  return Math.abs(a - b) <= eps * Math.max(1, Math.abs(a), Math.abs(b))
}

// ──────────────────────────────────────────────────────────────────────────────
// Default export
// ──────────────────────────────────────────────────────────────────────────────

export default {
  buildHashShare,
  verifyHashShare,
  drawU,
  score,
  computeNullifier,
  toEnvelopeCBOR,
  fromEnvelopeCBOR
}

File: sdk/rust/src/proofs/hashshare.rs
//! HashShare local build & verify helpers (developer tools).
//!
//! This module mirrors the reference "u-draw" used by miners to produce
//! HashShare proofs. It does **not** implement the full on-chain envelope
//! from `proofs/` — instead it lets SDK users and tests:
//!   * deterministically derive the uniform draw `u ∈ (0,1)`
//!   * compute H(u) = -ln(u) in micro-nats (µ-nats)
//!   * compare against a micro-threshold Θ (Theta) to decide acceptance
//!   * optionally scan a nonce range to find an acceptable share
//!
//! Hashing domain:
//!   SHA3-256( DOMAIN || header_hash(32B) || nonce_le(8B) || mix_seed? )
//! where `DOMAIN = b"anim/hs/v1"` and `mix_seed` is optional bytes used
//! by some networks to bind additional entropy (e.g., a mix digest).
//!
//! The uniform `u` is derived from the first 16 bytes (big-endian) of the
//! digest: `u = (v + 1) / 2^128`, ensuring `u ∈ (0,1]` (never zero).

use crate::error::{Error, Result};
use hex::FromHex;
use serde::{Deserialize, Serialize};
use sha3::{Digest, Sha3_256};

const DOMAIN: &[u8] = b"anim/hs/v1";
const TWO_POW_128: f64 = 340282366920938463463374607431768211456.0; // 2^128 as f64

/// Inputs to a local HashShare draw.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HashShareInput {
    /// 0x-hex, 32 bytes (block header sign-domain hash).
    pub header_hash: String,
    /// 64-bit nonce interpreted as little-endian when hashed.
    pub nonce: u64,
    /// Optional 0x-hex seed mixed into the domain (e.g., mix digest).
    #[serde(default)]
    pub mix_seed: Option<String>,
}

/// Result of a local HashShare draw.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HashShareDraw {
    /// Full SHA3-256 digest (0x-hex) of the draw preimage.
    pub digest: String,
    /// Uniform draw in (0,1]; represented as f64 for convenience.
    pub u: f64,
    /// H(u) in micro-nats (µ-nats) = round(-ln(u) * 1e6).
    pub h_micro: u64,
}

/// Verification summary.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerifyResult {
    /// Whether H(u) ≥ Θ (micro-nats).
    pub accepted: bool,
    /// H(u) in micro-nats.
    pub h_micro: u64,
    /// Difficulty ratio: H(u) / Θ (>= 1.0 implies acceptance).
    pub d_ratio: f64,
    /// The underlying draw (digest and u).
    pub draw: HashShareDraw,
}

/// Perform a local draw from inputs.
pub fn draw(input: &HashShareInput) -> Result<HashShareDraw> {
    let header = hex32(&input.header_hash)?;
    let mut hasher = Sha3_256::new();
    hasher.update(DOMAIN);
    hasher.update(header);

    // Nonce little-endian (matches miner inner-loop conventions).
    let nonce_le = input.nonce.to_le_bytes();
    hasher.update(nonce_le);

    if let Some(ms) = &input.mix_seed {
        let mix = hex_decode(ms)?;
        hasher.update(&mix);
    }

    let digest = hasher.finalize();
    let digest_hex = hex0x(&digest);

    // Uniform u from first 16 bytes (big-endian) mapped into (0,1].
    let mut hi16 = [0u8; 16];
    hi16.copy_from_slice(&digest[..16]);
    let v = u128::from_be_bytes(hi16);
    // Avoid zero for -ln(u): map v=0 → 1 (smallest non-zero) to cap H(u).
    let v = if v == 0 { 1 } else { v };
    let u = (v as f64) / TWO_POW_128;
    let h_micro = (-u.ln() * 1_000_000.0).round() as u64;

    Ok(HashShareDraw {
        digest: digest_hex,
        u,
        h_micro,
    })
}

/// Verify a draw against a micro-threshold Θ (Theta).
pub fn verify(input: &HashShareInput, theta_micro: u64) -> Result<VerifyResult> {
    let d = draw(input)?;
    let d_ratio = if theta_micro == 0 {
        f64::INFINITY
    } else {
        (d.h_micro as f64) / (theta_micro as f64)
    };
    let accepted = d_ratio >= 1.0;
    Ok(VerifyResult {
        accepted,
        h_micro: d.h_micro,
        d_ratio,
        draw: d,
    })
}

/// Attempt to find an acceptable share by scanning a bounded number of nonces.
/// Returns `Ok(Some((nonce, result)))` on success, `Ok(None)` if not found.
pub fn scan_for_share(
    mut input: HashShareInput,
    theta_micro: u64,
    max_tries: u64,
) -> Result<Option<(u64, VerifyResult)>> {
    for _ in 0..max_tries {
        let vr = verify(&input, theta_micro)?;
        if vr.accepted {
            return Ok(Some((input.nonce, vr)));
        }
        input.nonce = input.nonce.wrapping_add(1);
    }
    Ok(None)
}

// ------------------------------- Helpers ------------------------------------

fn hex0x(bytes: impl AsRef<[u8]>) -> String {
    format!("0x{}", hex::encode(bytes))
}

fn hex_decode(s: &str) -> Result<Vec<u8>> {
    let s = s.strip_prefix("0x").unwrap_or(s);
    Ok(Vec::<u8>::from_hex(s).map_err(|e| Error::InvalidData(format!("hex decode: {e}")))?)
}

fn hex32(s: &str) -> Result<[u8; 32]> {
    let v = hex_decode(s)?;
    if v.len() != 32 {
        return Err(Error::InvalidData(format!(
            "expected 32-byte hex, got {} bytes",
            v.len()
        )));
    }
    let mut out = [0u8; 32];
    out.copy_from_slice(&v);
    Ok(out)
}

// --------------------------------- Tests ------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn draw_is_deterministic() {
        let input = HashShareInput {
            header_hash: "0x000102030405060708090a0b0c0d0e0f\
                          101112131415161718191a1b1c1d1e1f"
                .to_string(),
            nonce: 42,
            mix_seed: Some("0xdeadbeef".to_string()),
        };
        let d1 = draw(&input).unwrap();
        let d2 = draw(&input).unwrap();
        assert_eq!(d1.digest, d2.digest);
        assert!((d1.u - d2.u).abs() < 1e-18);
        assert_eq!(d1.h_micro, d2.h_micro);
        assert!(d1.u > 0.0 && d1.u <= 1.0);
    }

    #[test]
    fn verify_ratio_behaves() {
        let input = HashShareInput {
            header_hash: "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa".into(),
            nonce: 1,
            mix_seed: None,
        };
        let vr = verify(&input, 1).unwrap();
        // Any non-zero H(u) over Θ=1 µ-nat should almost always be accepted;
        // but to avoid flakiness, check ratio is computed consistently.
        assert!(vr.d_ratio >= 0.0);
        assert_eq!(vr.accepted, vr.d_ratio >= 1.0);
    }

    #[test]
    fn scan_finds_or_not() {
        // Low Θ should be easy to satisfy.
        let mut found = false;
        let input = HashShareInput {
            header_hash: "0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb".into(),
            nonce: 0,
            mix_seed: None,
        };
        if let Ok(Some((_n, vr))) = scan_for_share(input, 1, 1000) {
            found = vr.accepted;
        }
        assert!(found);

        // Extremely high Θ likely won't be met in a tiny window.
        let input2 = HashShareInput {
            header_hash: "0xcccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc".into(),
            nonce: 0,
            mix_seed: None,
        };
        let none = scan_for_share(input2, 5_000_000_000, 128).unwrap();
        assert!(none.is_none());
    }

    #[test]
    fn hex_guards() {
        let bad = HashShareInput {
            header_hash: "0x1234".into(), // too short
            nonce: 0,
            mix_seed: None,
        };
        assert!(draw(&bad).is_err());
    }
}
